<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Compare dp mechanisms : Compare the performance of different differentially private mechanisms for releasing the most significant SNPs.">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Compare dp mechanisms</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/fy/compare_dp_mechanisms">View on GitHub</a>

          <h1 id="project_title">Compare dp mechanisms</h1>
          <h2 id="project_tagline">Compare the performance of different differentially private mechanisms for releasing the most significant SNPs.</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/fy/compare_dp_mechanisms/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/fy/compare_dp_mechanisms/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h1>
<a id="introduction" class="anchor" href="#introduction" aria-hidden="true"><span class="octicon octicon-link"></span></a>Introduction</h1>

<p>Code in this repository compare the performace of three differentially private mechanisms: the Laplace mechanism, the exponential mechanism, and the method proposed by <a href="http://dl.acm.org/citation.cfm?id=2487687">Johnson &amp; Shmatikov (2013)</a>. The data used for demonstration are simulated genome-wide association study data generated by <a href="http://www.hapsample.org/">HAP-SAMPLE</a>.</p>

<p>Code in this repository were used to produce results in the following paper:</p>

<blockquote>
<p>"Scalable privacy-preserving data sharing methodology for genome-wide association studies." Yu, F, S. Fienberg, S. SlavkoviÄ‡, C. Uhler (2014). Journal of Biomedical Informatics (forthcoming).  </p>
</blockquote>

<p>The paper is available at <a href="http://dx.doi.org/10.1016/j.jbi.2014.01.008">here</a> and <a href="http://arxiv.org/abs/1401.5193">here</a>.</p>

<h1>
<a id="example-data-generation" class="anchor" href="#example-data-generation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Example data generation</h1>

<p>The example case/control genotype data (<code>example/case_genotypes.dat</code> and <code>example/case_genotypes.dat</code>) were generated by <a href="http://www.hapsample.org/">HAP-SAMPLE</a> with the following options:</p>

<ul>
<li>Population: CEU</li>
<li>Source for SNPs: <code>Chrom9_Chrom13_snps.txt</code> </li>
<li>Disease Model File: <code>AR_chrom9_chrom13_Nov09_additive1_MAF025.txt</code>
</li>
<li>Simulation Type: Case/Control</li>
<li>Number of Cases: 1000</li>
<li>Number of Controls: 1000</li>
<li>Average breaks per cM: 1</li>
<li>Output Format: SNPs v. Individuals</li>
</ul>

<p>The disease model file describes a disease with 2 causative SNPs having addive effects. For more details about the disease model, see <a href="http://arxiv.org/abs/1006.4929">Malaspinas &amp; Uhler (2010)</a>.</p>

<h1>
<a id="analysis-on-the-example-data" class="anchor" href="#analysis-on-the-example-data" aria-hidden="true"><span class="octicon octicon-link"></span></a>Analysis on the example data</h1>

<p>See <a href="http://nbviewer.ipython.org/github/fy/compare_dp_mechanisms/blob/master/notebooks/compare_all_dp_mechanisms.ipynb">http://nbviewer.ipython.org/github/fy/compare_dp_mechanisms/blob/master/notebooks/compare_all_dp_mechanisms.ipynb</a></p>

<p>Or if you have ipython notebook installed, run in the terminal</p>

<pre><code>cd notebooks
../start_ipynb.sh
</code></pre>

<p>and open the <code>compare_all_dp_mechanisms</code> notebook in the browser. By default, the iPyhon notebook is at <a href="http://localhost:8888/">http://localhost:8888/</a></p>

<h1>
<a id="how-to-use-the-code" class="anchor" href="#how-to-use-the-code" aria-hidden="true"><span class="octicon octicon-link"></span></a>How to use the code</h1>

<p>First of all, untar the example data:</p>

<pre><code>tar xvfz exmaple.tar.gz
</code></pre>

<p>Then convert raw data to genotype tables:</p>

<pre><code>python ./notebooks/raw_to_geno_table.py ./example/case_genotypes.dat ./example/anticase_genotypes.dat ./table.tmp
</code></pre>

<p>To get results from the <strong>Laplace mechanism</strong>, run:</p>

<pre><code>python ./notebooks/write_chisquare.py ./table.tmp ./chisquare.tmp
python ./notebooks/get_laplace_results.py k e n_case n_control ./chisquare.tmp
</code></pre>

<p>where "k" is the number of top SNPs to release, "e" is the privacy budget (commonly known as epsilon in epsilon-differential privacy), and "n_case" and "n_control" are the numbers of cases and controls.</p>

<p>Similarly, to get results from the <strong>Exponential mechanism</strong>, run:</p>

<pre><code>python ./notebooks/write_chisquare.py ./table.tmp ./chisquare.tmp
python ./notebooks/get_expo_results.py k e n_case n_control ./chisquare.tmp
</code></pre>

<p>To get results from the <strong>Johnson &amp; Shmatikov method</strong>, run:</p>

<pre><code>python ./notebooks/write_JS_distance.py -p pval ./table.tmp ./js_distance.tmp
python ./notebooks/get_JS_results.py k e ./js_distance.tmp
</code></pre>

<p>where "pval" is the p-value specified in <a href="http://dl.acm.org/citation.cfm?id=2487687">Johnson &amp; Shmatikov (2013)</a>, which can be interpreted as the overall p-value (say 0.05) of the multiple testing problem involving thousands of SNPs.</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Compare dp mechanisms maintained by <a href="https://github.com/fy">fy</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

              <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-63494301-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>


  </body>
</html>
