# Introduction #

Code in this repository compare the performace of three differentially private mechanisms: the Laplace mechanism, the exponential mechanism, and the method proposed by [Johnson & Shmatikov (2013)](http://dl.acm.org/citation.cfm?id=2487687). The data used for demonstration are simulated genome-wide association study data generated by [HAP-SAMPLE](http://www.hapsample.org/).


Code in this repository were used to produce results in the following paper:

> "Scalable privacy-preserving data sharing methodology for genome-wide association studies." Yu, F, S. Fienberg, S. SlavkoviÄ‡, C. Uhler (2014). Journal of Biomedical Informatics (forthcoming).  

The paper is available at [here](http://dx.doi.org/10.1016/j.jbi.2014.01.008) and [here](http://arxiv.org/abs/1401.5193).





# Example data generation #

The example case/control genotype data (`example/case_genotypes.dat` and `example/case_genotypes.dat`) were generated by [HAP-SAMPLE](http://www.hapsample.org/) with the following options:

* Population: CEU
* Source for SNPs: `Chrom9_Chrom13_snps.txt` 
* Disease Model File: `AR_chrom9_chrom13_Nov09_additive1_MAF025.txt`
* Simulation Type: Case/Control
* Number of Cases: 1000
* Number of Controls: 1000
* Average breaks per cM: 1
* Output Format: SNPs v. Individuals

The disease model file describes a disease with 2 causative SNPs having addive effects. For more details about the disease model, see [Malaspinas & Uhler (2010)](http://arxiv.org/abs/1006.4929).



# Analysis on the example data #
See http://nbviewer.ipython.org/github/fy/compare_dp_mechanisms/blob/master/notebooks/compare_all_dp_mechanisms.ipynb

Or if you have ipython notebook installed, run in the terminal

    ./start_ipynb.sh

and open the `compare_all_dp_mechanisms` notebook in the browser. 


# How to use the code #
First of all, untar the example data:

    tar xvfz exmaple.tar.gz

Then convert raw data to genotype tables:

    python ./notebook/raw_to_geno_table.py ./example/case_genotypes.dat ./example/anticase_genotypes.dat ./table.tmp

To get results from the **Laplace mechanism**, run:

    python ./notebook/write_chisquare.py ./table.tmp ./chisquare.tmp
    python ./notebook/get_laplace_results.py k e n_case n_control ./chisquare.tmp

where "k" is the number of top SNPs to release, "e" is the privacy budget (commonly known as epsilon in epsilon-differential privacy), and "n_case" and "n_control" are the numbers of cases and controls.

Similarly, to get results from the **Exponential mechanism**, run:

    python ./notebook/write_chisquare.py ./table.tmp ./chisquare.tmp
    python ./notebook/get_expo_results.py k e n_case n_control ./chisquare.tmp

To get results from the **Johnson & Shmatikov method**, run:

    python ./notebook/write_JS_distance.py -p pval ./table.tmp ./js_distance.tmp
    python ./notebook/get_JS_results.py k e ./js_distance.tmp

where "pval" is the p-value specified in [Johnson & Shmatikov (2013)](http://dl.acm.org/citation.cfm?id=2487687), which can be interpreted as the overall p-value (say 0.05) of the multiple testing problem involving thousands of SNPs.
